infra/gcp/deploy-cip-auditor.sh

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env bash
+#
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO(listx): CONVERT THIS TO TERRAFORM!
+
+# Deploy Cloud Run service named "cip-auditor".
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+CLOUD_RUN_SERVICE_NAME="cip-auditor"
+SUBSCRIPTION_NAME="cip-auditor-invoker"
+SERVICE_ACCOUNT="k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com"
+
+# This sets up the GCP project so that it can be ready to deploy the cip-auditor
+# service onto Cloud Run.
+this::prepare_env()
+{
+    # Enable APIs.
+    services=(
+		"serviceusage.googleapis.com"
+		"cloudresourcemanager.googleapis.com"
+		"stackdriver.googleapis.com"
+		"clouderrorreporting.googleapis.com"
+		"run.googleapis.com"
+    )
+    for service in "${services[@]}"; do
+        gcloud --quiet services enable "${service}" --project="${PROJECT_ID}"
+    done
+
+    # Create "gcr" topic if it doesn't exist yet.
+	if ! gcloud pubsub topics list --format='value(name)' --project="${PROJECT_ID}" \
+        | grep "projects/${PROJECT_ID}/topics/gcr"; then
+
+		gcloud pubsub topics create gcr --project="${PROJECT_ID}"
+    fi
+
+    # Allow the Pub/Sub to create auth tokens in the project. This is part of
+    # the authentication bridge between the "gcr" Pub/Sub topic and the
+    # "--no-allow-unauthenticated" Cloud Run service option.
+    gcloud \
+        projects \
+        add-iam-policy-binding \
+        "${PROJECT_ID}" \
+        "--member=serviceAccount:service-${PROJECT_NUMBER}@gcp-sa-pubsub.iam.gserviceaccount.com" \
+        "--role=roles/iam.serviceAccountTokenCreator"
+}
+
+this::deploy_cip_auditor()
+{
+    # Deploy the auditor. The "--update-env-vars" is critical as it points to
+    # the reference set of official promoter manifests to validate GCR changes
+    # against. Note that there is no need to specify the
+    # [{asia,eu,us}.]gcr.io/k8s-artifacts-prod GCR (which we want to audit),
+    # because the "gcr" Pub/Sub topic on the project will automatically pick up
+    # changes seen in [{asia,eu,us}.]gcr.io/k8s-artifacts-prod.
+    gcloud run deploy "${CLOUD_RUN_SERVICE_NAME}" \
```

Comment:
Is this idempotent?  If I run twice with the same digest, will it correctly elide one run?  If I run with a new digest will it kill the old version?

Reply:
> Is this idempotent? If I run twice with the same digest, will it correctly elide one run?
No. Even if we pass in the same args to gcloud (using the same image), this command will deploy a new Cloud Run "revision" and move traffic to it from the existing revision (each revision in Cloud Run has a UUID).

> If I run with a new digest will it kill the old version?

No, the old revision is still displayed in the Cloud Run dashboard, so you can switch back to it. However, running the `gcloud run deploy ...` command will make it so that the newly-deployed revision receives 100% of all traffic.

Reply:
Comment to this effect?  It's not idempotent but it's safe